# *✈️ AVIATION ANALYSIS PROJECT*
## *Assessing Aircraft Accident Trends and Safety Risks*

## 🎯 Project Overview  
This project analyzes **aircraft accident data (1962-2023)** to identify trends, risk factors, and insights into aviation safety. Using **Python (pandas, NumPy, Matplotlib)** and **Tableau**, we aim to:  

- 🛠️**Clean and preprocess** aviation accident data.  
- 🔍**Explore trends** in accident frequency, causes, and severity.  
- 📊**Visualize key insights** through charts and interactive dashboards.  
- 📉**Identify risk factors** that contribute to aviation incidents.  

## 📂 Dataset Information  
- **`'AviationData.csv'`** – Contains accident reports, aircraft details, causes, and locations.  
- **`'USState_Codes.csv'`**– Mapping of state codes to full state names for analysis.  

## ❓ Key Questions  
- What are the most common causes of aviation accidents?  
- How have accident trends changed over time?  
- Which aircraft types and flight phases have the highest risk?  
- What role does weather or location play in accidents?  

**Let’s dive into the data!**  

## **Data Loading**
### **Data Loading: Importing the Dataset for Analysis**  

The first step in any data analysis process is loading the dataset into the environment. 
This allows us to inspect, clean, and prepare the data for further exploration.  


### **1. Importing Necessary Libraries**  
Before loading the datasets provided, essential **Python libraries** are imported for data manipulation and analysis.

### **2. Reading the Dataset**  
The datasets that are provided will then be loaded using **Pandas**.

### **3. Inspecting the Data**  
After loading, initial checks shall then be performed to understand the dataset structure:  

### **4. Verifying Data Integrity**  
To ensure data integrity, the following are checked for:  
- **Duplicate values** – using `.duplicated().sum()`  
- **Inconsistent column names** – using `aviation_df.columns`  
- **Missing values** – using `.isnull().sum()`  


In [79]:
# Import necessary libraries

import pandas as pd  # For data manipulation
import numpy as np  # For numerical operations
import matplotlib.pyplot as plt  # For data visualization
import seaborn as sns  # For statistical plots


In [80]:
# Loading the datasets as DataFrames
aviation_df = pd.read_csv("AviationData.csv", encoding="latin1", low_memory=False)  
state_codes_df = pd.read_csv("USState_Codes.csv", encoding="latin1")  


### Understanding the DataFrame  

The dataframe contains **aviation accident records** from 1962 to 2023. Each row represents an individual accident, including details such as:  

- **Accident Number** – A unique identifier for each incident.  
- **Date** – When the accident occurred.  
- **Location** – City and state of the accident.  
- **Aircraft Type** – The type of aircraft involved.  
- **Injuries and Fatalities** – Number of people affected.  
- **Cause of Accident** – Probable causes based on investigation reports.  

Additionally, we have a **USState_Codes.csv** file that maps **state abbreviations to full names** for location-based analysis.  

Let’s start by examining the dataframe structure!  

In [81]:
aviation_df.head()

,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
2,20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
3,20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
4,20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


In [82]:
aviation_df.shape

(88889, 31)

In [83]:
# Let's look at the concise summary of the data
aviation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

In [84]:
aviation_df.describe()

,Number.of.Engines,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured
count,82805.000000,77488.000000,76379.000000,76956.000000,82977.000000
mean,1.146585,0.647855,0.279881,0.357061,5.325440
std,0.446510,5.485960,1.544084,2.235625,27.913634
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,1.000000
75%,1.000000,0.000000,0.000000,0.000000,2.000000
max,8.000000,349.000000,161.000000,380.000000,699.000000


### DataFrame Summary  

The `.info()` method gives us an overview of the dataframe, including:  

- **Total entries**: Number of rows in the dataframe.  
- **Column names & data types**: Helps us understand which columns are numerical, categorical, or text-based.  
- **Missing values**: Identifies columns with missing data that may require cleaning.  

### Key Observations  
1. The data has **88889 rows** and **31 columns**.
2. There are 5 columns that have **Numeric Data(floats)** and 26 columns that have **Categorical Data(objects)**.
3. 29 columns out of the 31 have **missing values**.

### Statistical Summary of the DataFrame

The `.describe()` function provides key summary statistics for **numerical columns**, helping us understand:  

- **Count** – Number of non-missing values.  
- **Mean** – Average value for each column.  
- **Standard Deviation** – How spread out the values are.  
- **Min & Max** – The range of values.  
- **Quartiles (25%, 50%, 75%)** – Helps identify distribution patterns.  

### Key Insights  
1. Identify **outliers** by looking at min/max values.  
2. Check for **skewed distributions** using mean vs. median (50%).  
3. Compare **data spread** using standard deviation.  

In [85]:
aviation_df.duplicated().sum()

0

In [86]:
#Identify missing values
aviation_df.isnull().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38640
Airport.Name              36099
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1317
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6381
Publication.Date          13771
dtype: i

## Data Cleaning
### **Preparing the Dataset for Analysis**  

Data cleaning is a crucial step in the data analysis process. It ensures that our dataset is accurate, consistent, and ready for further analysis. Below are the key steps performed to clean our aviation accident dataset: 

### **1. Standardizing Column Names**  
To ensure consistency in the dataset, the following was done:  
- Capitalizing the first letter of every word in column names.  


### **2. Handling Missing Values**  
Missing values can affect the reliability of our analysis. Different strategies were used to handle them:  

- **Numerical Data:** Filled missing values with the median, rounding up to the nearest whole number.  
- **Categorical Data:** Filled missing values with the most frequent value (mode).  
- **Columns with High Missing Percentages:** Considered removing features with excessive missing data that provided little value.  
 

### **3. Adding New Features for Better Insights**  
New features were added to enhance our analysis:  


### **Outcome: A Clean & Structured Dataset**  
After these steps, the dataset is now:  
- Free of missing values.  
- Consistent in format.  
- Enhanced with new features for deeper analysis.  


In [87]:
# Standardizing column names: Replace dots with underscores and capitalize first letter of each word
aviation_df.columns = aviation_df.columns.str.replace('.', '_').str.title()

# Display updated column names
aviation_df.columns

Index(['Event_Id', 'Investigation_Type', 'Accident_Number', 'Event_Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport_Code',
       'Airport_Name', 'Injury_Severity', 'Aircraft_Damage',
       'Aircraft_Category', 'Registration_Number', 'Make', 'Model',
       'Amateur_Built', 'Number_Of_Engines', 'Engine_Type', 'Far_Description',
       'Schedule', 'Purpose_Of_Flight', 'Air_Carrier', 'Total_Fatal_Injuries',
       'Total_Serious_Injuries', 'Total_Minor_Injuries', 'Total_Uninjured',
       'Weather_Condition', 'Broad_Phase_Of_Flight', 'Report_Status',
       'Publication_Date'],
      dtype='object')

This ensures that the column names have the same format.

### Handling Missing Data  

Missing values can impact analysis and visualization accuracy. We need to:  

1. **Identify columns with missing values** – Check how many and where they appear.  
2. **Decide on a handling strategy**:  
   - **Fill missing values** using appropriate methods (e.g., mean, median, mode).  
   - **Drop columns or rows** if missing values are excessive.  
   - **Use domain knowledge** to make informed replacements.  

Let's first analyze the extent of missing data before deciding the best approach!

In [88]:
# Checking for missing values in the dataset
missing_values = aviation_df.isnull().sum()

# Displaying only columns with missing values
missing_values = missing_values[missing_values > 0]

# Checking percentage of missing values
missing_percentage = (missing_values / len(aviation_df)) * 100

# Display missing values with percentage
missing_data = pd.DataFrame({'Missing Values': missing_values, 'Percentage': missing_percentage})
missing_data

,Missing Values,Percentage
Location,52,0.058500
Country,226,0.254250
Latitude,54507,61.320298
Longitude,54516,61.330423
Airport_Code,38640,43.469946
Airport_Name,36099,40.611324
Injury_Severity,1000,1.124999
Aircraft_Damage,3194,3.593246
Aircraft_Category,56602,63.677170
Registration_Number,1317,1.481623


### Dropping Columns or Rows
- If a column has too many missing values (e.g., above 50-60%), we may drop it.
- If a column is less useful, it may not add much value to the analysis hence we drop it.
- If only a few rows are missing values in critical columns, we can drop them.

In [89]:
# Drop columns with more than 60% missing values
aviation_df = aviation_df.dropna(thresh=len(aviation_df) * 0.4, axis=1)

# Drop rows with any missing critical values
aviation_df = aviation_df.dropna(subset=['Make', 'Model', 'Amateur_Built'])


### Filling Missing Values
- For numerical columns, we can fill missing values with mean, median, or mode.
- For categorical columns, we can fill with mode (most common value) or 'Unknown'.

In [90]:
# Filling numerical columns with median
aviation_df['Total_Fatal_Injuries'] = aviation_df['Total_Fatal_Injuries'].fillna(np.ceil(aviation_df['Total_Fatal_Injuries'].mean()))
aviation_df['Total_Serious_Injuries'] = aviation_df['Total_Serious_Injuries'].fillna(np.ceil(aviation_df['Total_Serious_Injuries'].mean()))
aviation_df['Total_Minor_Injuries'] = aviation_df['Total_Minor_Injuries'].fillna(np.ceil(aviation_df['Total_Minor_Injuries'].mean()))
aviation_df['Total_Uninjured'] = aviation_df['Total_Uninjured'].fillna(round(aviation_df['Total_Uninjured'].mean()))
aviation_df['Number_Of_Engines'] = aviation_df['Number_Of_Engines'].fillna(aviation_df['Number_Of_Engines'].median())



In [91]:
# Selecting only categorical columns
categorical_cols = aviation_df.select_dtypes(include=['object'])

# Checking for missing values
missing_values_cat = categorical_cols.isnull().sum()

# Displaying only columns with missing values
missing_values_cat = missing_values_cat[missing_values_cat > 0]

# Checking percentage of missing values
missing_percentage_cat = (missing_values_cat / len(aviation_df)) * 100

# Display missing values with percentage
missing_data_cat = pd.DataFrame({'Missing Values': missing_values_cat, 'Percentage': missing_percentage_cat})
missing_data_cat



,Missing Values,Percentage
Location,48,0.054128
Country,225,0.253727
Airport_Code,38474,43.386184
Airport_Name,35942,40.530910
Injury_Severity,979,1.103994
Aircraft_Damage,3154,3.556688
Registration_Number,1208,1.362232
Engine_Type,6928,7.812535
Purpose_Of_Flight,6073,6.848373
Weather_Condition,4381,4.940346


### **Approach**
Categorize missing values into three levels:

1. **Low Missing Percentage (≤ 5%)**  
   - Filled using **Mode** (most frequent value in the column).  
   - Ensures minimal data distortion.

2. **Moderate Missing Percentage (5% - 20%)**  
   - Replaced with `"Unknown"` or `"Not Available"`.  
   - Helps preserve data integrity when missing values may hold meaning.

3. **High Missing Percentage (> 20%)**  
   - Uses **Forward Fill (ffill)** first, filling missing values with the previous row's value.  
   - If values are values are still missing, applies **Backward Fill (bfill)** to fill using the next row’s value.  

In [92]:
total_rows = len(aviation_df)

for col in missing_values_cat.index:
    missing_percentage_cat = (missing_values_cat[col] / total_rows) * 100  # Calculate missing percentage
    
    if missing_percentage_cat <= 5:  # If missing values are 5% or less
        aviation_df[col] = aviation_df[col].fillna(aviation_df[col].mode()[0])
    
    elif 5 < missing_percentage_cat <= 20:  # If missing values are between 5% and 20%
        aviation_df[col] = aviation_df[col].fillna("Unknown")
    
    else:  # If missing values are more than 20%
        aviation_df[col] = aviation_df[col].fillna(method='ffill')  # First, try forward fill
        aviation_df[col] = aviation_df[col].fillna(method='bfill')  # Then, try backward fill if still missing

# Confirm missing values are handled
print("Missing values after handling:")
print(aviation_df.isnull().sum())

Missing values after handling:
Event_Id                  0
Investigation_Type        0
Accident_Number           0
Event_Date                0
Location                  0
Country                   0
Airport_Code              0
Airport_Name              0
Injury_Severity           0
Aircraft_Damage           0
Registration_Number       0
Make                      0
Model                     0
Amateur_Built             0
Number_Of_Engines         0
Engine_Type               0
Purpose_Of_Flight         0
Total_Fatal_Injuries      0
Total_Serious_Injuries    0
Total_Minor_Injuries      0
Total_Uninjured           0
Weather_Condition         0
Broad_Phase_Of_Flight     0
Report_Status             0
Publication_Date          0
dtype: int64


Now that the data is filled with no none or null values we can proceed with coverting the data in the Event_Date column to the correct format.

In [93]:
# Convert Event_Date to datetime format
aviation_df['Event_Date'] = pd.to_datetime(aviation_df['Event_Date'])


### Extracting Time-Based Features
Since accidents often have seasonal trends, let’s create features from the date column:

In [94]:
# Extracting key time-related features
aviation_df['Year'] = aviation_df['Event_Date'].dt.year
aviation_df['Month'] = aviation_df['Event_Date'].dt.month
aviation_df['Day_Of_Week'] = aviation_df['Event_Date'].dt.day_name()

The above will:
- Help to analyze if accidents occur more frequently on weekends.
- Detect the seasonal trends (e.g., winter vs. summer accidents).

### Counting Incidents  
To better understand aviation accident trends, create new features that count the number of incidents based on different categories.  



In [95]:
aviation_df['Manufacturer_Incident_Count'] = aviation_df.groupby('Make')['Make'].transform('count') # Count incidents per aircraft manufacturer (Make)
aviation_df['Model_Incident_Count'] = aviation_df.groupby('Model')['Model'].transform('count') # Count incidents per aircraft model
aviation_df['Flight_Purpose_Incident_Count'] = aviation_df.groupby('Purpose_Of_Flight')['Purpose_Of_Flight'].transform('count') # Count incidents per flight purpose
aviation_df['Weather_Incident_Count'] = aviation_df.groupby('Weather_Condition')['Weather_Condition'].transform('count') # Count incidents per weather condition



**groupby('Make')** groups the dataset by `Make`:
- Counts how many times each manufacturer appears in the dataset (i.e., how many accidents involve aircraft from that manufacturer).
- Assigns the count as a new column manufacturer_incident_count for each row.

**groupby('Model')** groups the dataset by `Model`:
- Counts how many accidents occurred for each model.
- Stores the count in model_incident_count.

**groupby('Purpose_Of_Flight')** groups accidents based on the `Purpose_Of_Flight`:
- Counts how many accidents occurred for each type of flight.
- Creates flight_purpose_incident_count to store the count.

**groupby('Weather_Condition')** groups accidents by `Weather_Condition`:
- Counts how many accidents happened under each weather type.
- Stores this in weather_incident_count to help analyze how weather impacts aviation safety.

By counting the number of incidents based on different categories, we are able to:
- Identify high-risk aircraft manufacturers and models.
- Detect which types of flights have more accidents.
- Analyze weather conditions associated with aviation incidents.
- Improve accident prediction models by adding meaningful features.

### Identify Key Features

Categorize the features into potentially useful and less useful based on their relevance to accident analysis.

The potentially useful features are to be kept and the less useful features are to be dropped.

Potentially useful features to be kept:
- `Event_Date`: Helps analyze accident trends over time.
- `Location`: Geographic patterns of accidents.
- `Airport_Name`: Identifying high-risk airports.
- `Injury_Severity`: The severity of accidents.
- `Aircraft_Damage`: Understanding the extent of damage.
- `Make` & `Model`: Aircraft manufacturers & models with the most accidents.
- `Number_Of_Engines` & `Engine_Type`: Engine configuration influence on accidents.
- `Purpose_Of_Flight`: Comparing accidents between commercial & private flights.
- `Total_Fatal_Injuries`, `Total_Serious_Injuries`, `Total_Minor_Injuries`, `Total_Uninjured`: Evaluate the accident impact.
- `Weather_Condition`: Role of weather in accidents.
- `Broad_Phase_Of_Flight`: Identifies when accidents occur (e.g., takeoff, landing).
- `Year`,`Month`,`Day_Of_Week`,`Manufacturer_Incident_Count`,`Model_Incident_Count`,`Flight_Purpose_Incident_Count`&`Weather_Incident_Count`: New features for deeper insights.

Less useful features to be dropped:
- `Airport_Code`: May not be necessary if we have Airport_Name.
- `Registration_Number`: Unique to each aircraft, not relevant for trend analysis.
- `Accident_Number`: Identifier that doesn’t add analytical value.
- `Report_Status`&`Publication_Date`: Administrative data, not directly useful for insights
- `Country`: The data is only for USA.


In [96]:
# Dropping irrelevant columns
columns_to_drop = ['Accident_Number', 'Registration_Number', 'Airport_Code', 'Report_Status', 'Publication_Date','Country']
aviation_df = aviation_df.drop(columns=columns_to_drop)


In [97]:
# Save Cleaned DataFrame to Excel
aviation_df.to_excel('Cleaned_Aviation_Data.xlsx', index=False)